### ISSUE TITLE
Categories default order looks wrong

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#2607]
Checked on latest trunk (after #2606), the default order of the categories in the Categories & Tags panel doesn't seem right.

I have a decent amount of categories in my local install (the ones from the Theme unit test data xml and a few more) so the list is pretty long. Comparing the order in the classic editor and in Gutenberg, they're different:

<img width="1278" alt="screen shot 2017-08-30 at 14 26 04" src="https://user-images.githubusercontent.com/1682452/29872853-86d79b62-8d91-11e7-95ce-f5c80bc36b81.png">

On the left: the classic editor list (I've avoided to expand the list to show all the categories)
On the right: the Gutenberg list split in 3 parts

In a fist time I thought it's an issue related to lowercase/uppercase first letter, but then noticed there's no apparent logic in the misplacement. Some categories with uppercase first letter have a wrong order, others don't.

From a user experience perspective, users should be able to quickly and easily find the category they're interested in and the current core implementation provides an alphabetical order, case insensitive.

**Comment by aduth** [comment]
The Gutenberg category picker sorts by "count" (occurrence of categories):

https://github.com/WordPress/gutenberg/blob/32647cd2075c8549d39c6982c7742270f584792e/editor/sidebar/post-taxonomies/hierarchical-term-selector.js#L19-L23

The current editor sorts by "name":

https://github.com/WordPress/WordPress/blob/4.8.1/wp-admin/includes/template.php#L227-L262

Agree that it's not intuitive if expecting to scan by names (alphabetical works better), but also surfaces common categories. Flagging for design, question.

**Comment by GaryJones** [comment]
How about pulling the top ~5 or so to the top per count, then a horizontal divider, and then the rest of the categories in name order?

**Comment by afercia** [comment]
Yep, maybe giving users the ability to scan by both (a-z and count) would be best.

From an accessibility perspective, this categories list has always been a bit problematic: when using the keyboard and there are dozens of categories, having to tab through all of them is really annoying. It would be great to have a way to "skip" them, but this is certainly out of the scope of this issue.

**Comment by bobbingwide** [comment]
Unless someone can provide a very good reason why the Category display should be changed then I see no reason why the display for the Categories and hierarchical custom categories should be any different from the current solution. 

Selected Categories first followed by the unselected Categories sorted alphabetically
![image](https://user-images.githubusercontent.com/2474435/38173325-7ff04cd8-35b4-11e8-9715-7c53109e2bc2.png)

Using Gutenberg, if you're unaware of the Most used context, the order is a complete mess.
![image](https://user-images.githubusercontent.com/2474435/38173390-b7029130-35b5-11e8-8150-0f68803b9c5f.png)


Also someone needs to explain 
- why the Most Used tab has been eliminated.
- why there's no scroll bar

The Parent select list for the Create new category action should also be ordered the same way.

This is not an enhancement. It's required to fix an undocumented backward incompatible change..

**Comment by kraftbj** [comment]
Personally, `name` is what I expected to. Working on a site that has been active for 16 years (pre-WP, migrated in at some point), so there is a lot of content over the years. New categories have been added, focus of the site shifted over the years, etc. The `count` makes it confusing and difficult to find some of the newer categories used on the site currently and presently "active".

With a name order, there is an obvious sort order and awareness of where to look for a particular category.

**Comment by TwisterMc** [comment]
I thought this was a bug not a feature. Please organize categories by name. I agree with everything @kraftbj is stating. I have 52 categories going back to 2004 and the category listing is just a mess right now.

If we're going to keep the current order, then put post counts next to the category name to help people understand why the order is what it is. Right now it seems random to new users.

**Comment by iandunn** [comment]
This confused me too. I agree that surfacing popular categories is a good thing, but I think that the side-effect of making all other categories difficult creates a problem that's much worse.

If they're alphabetical, then it's obvious and intuitive. Maybe it takes half a second to scan the list, but that's a lot less painful than having to look through every single item, especially when there are dozens, or even hundreds, of items.

Maybe a better way to solve the problem would be to add a search or autocomplete option instead?

**Comment by hlrnet** [comment]
Hi,
This default nuimbering by count is horrible. I have lots of categories and subcategories. It takes me ages to find a category I'm looking for.
If you can't provide a way to turn this 'feature' off, I'll be forced to switch Gutenberg off.

**Comment by notnownikki** [comment]
Hi folks, I'm starting to work on this today.

**Comment by straightvisions** [comment]
Don't see a way to filter sorting with the current code, as sorting is defined as non-filterable constant here:

https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

`const DEFAULT_QUERY = {
	per_page: -1,
	orderby: 'count',
	order: 'desc',
	_fields: 'id,name,parent',
};`

Sort by name should be default and custom filters should be supported.

**Comment by StaggerLeee** [comment]
I myself thought it is a bug. Have on one website 76 categores and subcategories, and it is mild to say it is big mess.

**Comment by chezmat** [comment]
Hi,

I was looking for a way to correct this and discover this topic. 
It's completly unbelievable that such a simple thing is open for more than one year!! 
It's stupid to sort by count. 
I have a hierarchical custom taxonomy, it's such a mess to find the good one in all the listing! 
@karmatosed your filter idea was really good, why did you left? :-)
So please, just make the rollback to previous alphabetical sorting asap. And by asap, i mean before version 5.0 of wordpress and mandatory gutenberg..

**Comment by karmatosed** [comment]
@chezmat I removed my assignment as it had a design, next up we get development so this is still potentially to work on.

**Comment by notnownikki** [comment]
If we're ok going with the design in https://github.com/WordPress/gutenberg/issues/2607#issuecomment-412651062 I can start looking at that? Although it doesn't seem to have the hierarchical view of categories that it used to have. Is that intentional?

**Comment by chezmat** [comment]
yes please!

**Comment by bobbingwide** [comment]
> If we're ok going with the design in [#2607 (comment)](https://github.com/WordPress/gutenberg/issues/2607#issuecomment-412651062) I can start looking at that? Although it doesn't seem to have the hierarchical view of categories that it used to have. Is that intentional?

The hierarchical view of categories is required. Not having it will lead to confusion when there are multiple categories with the same name in different parts of the hierarchy. 
Can you not simply start reproducing the current implementation in WordPress core?

**Comment by afercia** [comment]
Re: the design in https://github.com/WordPress/gutenberg/issues/2607#issuecomment-412651062 can we please stop using placeholders as replacement for labels? This is something that has been discussed at length in various other issues and it's far from ideal for accessibility. There's also a core Trac ticket about it, with a pretty good amount of documentation and literature linked, see https://core.trac.wordpress.org/ticket/40331

With regards to the general issue, I'd tend to think the best option is to just have parity with the current implementation in core: alphabetical and with hierarchy.

**Comment by notnownikki** [comment]
> I'd tend to think the best option is to just have parity with the current implementation in core: alphabetical and with hierarchy.

I agree totally.

So, plan is to sort alphabetically, hierarchy with indentation, limit the height of the list so large lists don't make the page incredibly long, and have a search box at the top of the category that does as-you-type filtering.

**Comment by StaggerLeee** [comment]
And please, please make a solo-mode. When one header is clicked other is automatically closed.
 No need for all those sections in right sidebar to stick to our eyes all the time open.

- Click to expand categories section. See, do, what you need.
- Click on another section, say featured image.
- Categories section is automatically closed.

Only one visible all the time is of course first one.

**Comment by GaryJones** [comment]
> When one header is clicked other is automatically closed.

@StaggerLeee The _behaviour_ of the sidebar sections is not related to the _content appearance_ of the categories section.

Please find an existing ticket, or open one, to voice your opinions about the behaviour of the sidebar sections.

**Comment by notnownikki** [comment]
Little bit of progress...

![newcategoryui](https://user-images.githubusercontent.com/3314354/45956509-96fa6b80-c00a-11e8-979c-ec4f5ab5ac50.gif)

**Comment by bobbingwide** [comment]
@notnownikki How do you intend to deal with multiple results on the search?
e.g. having orange under marmalade as well as Jam.

**Comment by iandunn** [comment]
Those look great @notnownikki, kudos :)

**Comment by bobbingwide** [comment]
What's the plan for showing categories which are already selected, when there are more than 100 categories, and some of the selected categories may not be returned in the first 100, and they're not top level categories.

**Comment by chezmat** [comment]
@notnownikki : i love that! 
@bobbingwide you're right. I think the selected items should be top of the list, anyway, it's not very convenient..
![image](https://user-images.githubusercontent.com/42534744/46013534-e608d500-c0cc-11e8-8188-c0ed3de29ff4.png)

**Comment by notnownikki** [comment]
Thanks for the feedback everyone :)

I have some refactoring to do to fix accessibility issues, and then I'll get the selected categories showing at the top, and then I think we'll have something that works quite nicely!

**Comment by tofumatt** [comment]
#10138 fixed this, but the PR message missed the format to trigger GitHub to close the issue. This is closed, thanks @notnownikki! 👍

**Comment by straightvisions** [comment]
Unfortunately, there is still a Bug which results in first entry not being in alphabetical order.

Tested with WP 4.9.8 and Gutenberg 4.0. This bug occurs on both, add post and edit post screens.

![screenshot-dev paymentandbanking com-2018 10 18-09-43-50](https://user-images.githubusercontent.com/4689792/47138685-6086d800-d2ba-11e8-89fe-7d83125690ea.png)
![screenshot-cloud2 straightvisions com-8443-2018 10 18-09-42-37](https://user-images.githubusercontent.com/4689792/47138696-667cb900-d2ba-11e8-83ae-e35a7115c148.png)

**Comment by notnownikki** [comment]
@straightvisions I can't reproduce that, could you open an issue and let us know the steps to produce please?

**Comment by afercia** [comment]
Maybe one more case to check: I've added a `newone` category from the form below the list and it ended up in the middle of the list:

<img width="350" alt="screen shot 2018-10-18 at 10 07 03" src="https://user-images.githubusercontent.com/1682452/47140290-23bce000-d2be-11e8-84f1-d16671185be6.png">

In the Classic Editor, the newly added categories are temporarily displayed as first item in the list, so they're immediately visible.

Refreshing the page in Gutenberg displays `newone` in the correct order but seems other categories are not sorted correctly, see `antiquarianism`:

<img width="314" alt="screen shot 2018-10-18 at 10 09 28" src="https://user-images.githubusercontent.com/1682452/47140405-6bdc0280-d2be-11e8-8a99-378343a2f8f4.png">

Note: the list of categories in my install comes from the [themes unit test data xml](https://codex.wordpress.org/Theme_Unit_Test).
